feature_discovery.md

```diff
@@ -0,0 +1,158 @@
+
+Any agent may send another agent this message type at any time.
+Implementers of agents that intend to support dynamic relationships
+and rich features are *strongly* encouraged to implement support
+for this message, as it is likely to be among the first messages
+exchanged with a stranger.
+
+##### `disclose` Message Type
+
+A `discover-features/disclose` message looks like this:
+
+```json
+{
+  "@type": "https://didcomm.org/discover-features/1.0/disclose",
+  "thread_id": "yWd8wfYzhmuXX3hmLNaV5bVbAjbWaU",
+  "body":{
+      "protocols": [
+        {
+          "ptid": "https://didcomm.org/tictactoe/1.0",
+          "roles": ["player"]
+        }
+      ]
+  }  
+}
+```
+
+The `protocols` field is a JSON array of __protocol support descriptor__
+objects that match the query. Each descriptor has a `pid` that contains
+a protocol version (fully qualified message family identifier such as
+`https://didcomm.org/tictactoe/1.0`), plus a `roles`
+array that enumerates the roles the responding agent
+can play in the associated protocol.
+
+Response messages say, "Here are some protocols I support that matched
+your query, and some things I can do with each one."
+
+##### Sparse Responses
+
+Responses do not have to contain exhaustive detail. For example, the following
+response is probably just as good:
+
+```json
+{
+  "type": "https://didcomm.org/discover-features/1.0/disclose",
+  "thread_id": "yWd8wfYzhmuXX3hmLNaV5bVbAjbWaU",
+  "protocols": [
+    {"ptid": "https://didcomm.org/tictactoe/1.0"}
+  ]
+}
+```
+
+The reason why less detail probably suffices is that agents do not need to
+know everything about one another's implementations in order to start an
+interaction--usually the flow will organically reveal what's needed. For
+example, the `outcome` message in the `tictactoe` protocol isn't needed
+until the end, and is optional anyway. Alice can start a tictactoe game
+with Bob and will eventually see whether he has the right idea about
+`outcome` messages.
+
+The missing `roles` in this response does not say, "I support no roles
+in this protocol." It says, "I support the protocol but
+I'm providing no detail about specific roles."
+
+Even an empty `protocols` map does not say, "I support no protocols
+that match your query." It says, "I'm not telling you that I support any
+protocols that match your query." An agent might not tell another that
+it supports a protocol for various reasons, including: the trust that
+it imputes to the other party based on cumulative interactions so far,
+whether it's in the middle of upgrading a plugin, whether it's currently
+under high load, and so forth. And responses to a `discover-features` request are
+not guaranteed to be true forever; agents can be upgraded or downgraded,
+although they probably won't churn in their protocol support from moment
+to moment.
+
+### Privacy Considerations
+
+Because the regex in a `request` message can be very inclusive, the `discover-features`
```

Comment:
> regex

* we need to be very clear about the format of the contents of `query`. The concept of "regular expressions" appears once in this text, only here, and not in the section that defines `query`
* having said the previous - regular expressions seem like overkill... I forsee most implementations using just wildcards  on protocol names and versions... not sure about the authority... are extensions allowed under `didcomm.org`?

Reply:
I'd prefer that we not suggest regex as a method to process the different features. It opens up the possibility for [ReDoS attacks](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS) for implementers of this feature.

Reply:
It isn't actually a regex, just wildcards (*)

Reply:
Thanks for clarifying. That makes sense. 👍